### PR TITLE
Debug refactor

### PR DESCRIPTION
**Pull request by A2va**
This PR refactor the debugging part of this extension.
It adds an `xmake` debug type configuration, so you can have a configuration like this:
```json
{
    "configurations": [
    {
       "name": "XMake Debug",
        "type": "xmake",
        "request": "launch",
        "target": "example",
        "args":["-args"],
        "stopAtEntry": true
    }
  ]
}
``` 

### Changelog

* Add new debugger type, document it in the README. 
* Remove `runAll` in explorer, because it was not working with the previous versions, and I didn't manage to get
it works with the new debug configuration (maybe I'll get it to work in a future contribution).

## Config changes

* Remove `runningTargetsArguments`, `debuggingTargetsArguments` and `customDebugConfig`
* Rename `debugConfigType` to `debuggerBackend`
* Add `envBehaviour` config to manage environment between launch.json and xmake.

### CODE REVIEW

**Comment by waruqi** [comment]
I'm glad you're contributing to xmake, but I still don't see what this patch actually improves. And it breaks a lot of existing configurations, which may affect the experience of many users now.

> Remove runningTargetsArguments, debuggingTargetsArguments and customDebugConfig
Rename debugConfigType to debuggerBackend

We should not remove them. Even if there are new configuration interfaces, they should be compatible with the existing old configuration. Otherwise a large number of users will be affected.

> It adds an xmake debug type configuration, so you can have a configuration like this:

customDebugConfig does the same thing, so why add an extra configuration?

File: assets/target_informations.lua [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+function _get_path(target)
+	if target then
+		return path.join(os.projectdir(), target:targetfile())
+	end
+end
+
+function _get_rundir(target)
+	if target then
+		return path.absolute(target:rundir())
+	end
```
Comment (waruqi) [inline assets/target_informations.lua:34]: incorrect ident

File: assets/target_informations.lua [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+end
+
+function _get_envs(target)
+	local envs = {}
+    if target then
```
Comment (waruqi) [inline assets/target_informations.lua:39]: indent

File: assets/target_runenvs.lua [hunk 3]
```diff
@@ -64,20 +64,7 @@ function main (targetname)
...
-        empty = false
-    end
-    if json.mark_as_array then
-        json.mark_as_array(envirnoments)
-    end
-    if not empty then
-        print(json.encode(envirnoments))
-    else
-        print("[]")
-    end
```
Comment (waruqi) [inline assets/target_runenvs.lua:80]: why? we need fix empty array. otherwise it will be incorrect `{}`, not `[]`

File: package.json [hunk 4]
```diff
@@ -37,7 +38,6 @@
     "onCommand:xmake.onRebuild",
     "onCommand:xmake.onClean",
     "onCommand:xmake.onCleanAll",
-    "onCommand:xmake.onBuildRun",
```
Comment (waruqi) [inline package.json:40]: why remove it? Don't make destructive changes. Many other users need it.

File: package.json [hunk 5]
```diff
@@ -57,13 +57,6 @@
   ],
   "main": "./out/src/extension",
   "contributes": {
-    "keybindings": [
-      {
-        "key": "f5",
-        "command": "xmake.onDebug",
-        "when": "xmakeEnabled && debuggersAvailable && debugState == 'inactive'"
-      }
-    ],
```
Comment (waruqi) [inline package.json:66]: why?...

File: package.json [hunk 6]
```diff
@@ -146,11 +139,6 @@
         "title": "CleanAll",
         "category": "XMake"
       },
-      {
-        "command": "xmake.onBuildRun",
-        "title": "BuildRun",
-        "category": "XMake"
-      },
```
Comment (waruqi) [inline package.json:153]: why?

File: package.json [hunk 7]
```diff
@@ -425,6 +400,153 @@
...
+                "default": "integrated"
+              },
+              "env": {
+								"description": "Additional environment variables.",
+								"type": "object",
+								"patternProperties": {
+									".*": {
+										"type": "string"
+									}
+								},
```
Comment (waruqi) [inline package.json:462]: incorrect indent

File: package.json [hunk 8]
```diff
@@ -509,33 +631,24 @@
...
-            "default": []
-          },
-          "description": "The Running Targets Arguments, .e.g {\"targetName\": [\"args\", \"...\"]}"
-        },
-        "xmake.debuggingTargetsArguments": {
-          "type": "object",
-          "default": {
-            "default": []
-          },
-          "description": "The Debugging Targets Arguments, .e.g {\"targetName\": [\"args\", \"...\"]}"
```
Comment (waruqi) [inline package.json:524]: why? please revert it.

File: package.json [hunk 9]
```diff
@@ -509,33 +631,24 @@
...
+          "default":"merge",
+          "enum": ["merge", "override"],
+          "description": "Environment behaviour between launch.json and xmake envs", 
+          "enumDescriptions": [
+            "This will concat launch.json envs and xmake envs",
+            "Launch configurations will override xmake envs"
+          ]
         },
-        "xmake.debugConfigType": {
+        "xmake.debuggerBackend": {
```
Comment (waruqi) [inline package.json:644]: and here, This time breaking will the user's existing configuration.

File: package.json (hunk 8 above)
Comment (A2va) [reply package.json:524]: It can configured with args in launch.json

File: src/explorer.ts [hunk 10]
```diff
@@ -747,9 +747,9 @@ export class XMakeExplorer implements vscode.Disposable {
             }
         });
 
-        vscode.commands.registerCommand('xmakeExplorer.runAll', () => {
-            vscode.commands.executeCommand("xmake.onRun");
-        });
+        // vscode.commands.registerCommand('xmakeExplorer.runAll', () => {
+        //     vscode.commands.executeCommand("xmake.onRun");
+        // });
```
Comment (waruqi) [inline src/explorer.ts:752]: remove unused comments

File: src/process.ts [hunk 11]
```diff
@@ -20,8 +19,23 @@ export interface IExecutionHandle {
...
+    const iconv = require("iconv-lite");
+
+    const isWin = os.platform() === 'win32';
+    var charset = 'utf8';
+    if (isWin) {
+        charset = 'gbk';
+    }
+
+    return iconv.decode(bytes, charset).toString("utf8");
+}
```
Comment (waruqi) [inline src/process.ts:35]: it's common api, why move it?

File: src/xmake.ts [hunk 12]
```diff
@@ -648,106 +641,34 @@ export class XMake implements vscode.Disposable {
...
-            args.push(targetName);
-            for (let arg of configArgs) {
-                args.push(arg);
-            }
-        } else if (targetName == "all") {
-            args.push("-a");
-        } else {
-            for (let arg of configArgs) {
-                args.push(arg);
-            }
```
Comment (waruqi) [inline src/xmake.ts:745]: oh, why?

File: src/process.ts (hunk 11 above)
Comment (A2va) [reply src/process.ts:35]: It was used only in `process.ts`, so I moved it there.

File: src/xmake.ts (hunk 12 above)
Comment (A2va) [reply src/xmake.ts:745]: It can be configured with args in launch.json

File: src/process.ts (hunk 11 above)
Comment (waruqi) [reply src/process.ts:35]: It is a generic interface, with the possibility of adding other interfaces in the future.

I believe that each pr should make the smallest possible changes that are only relevant to the current functionality.

**Comment by A2va** [comment]
| We should not remove them. Even if there are new configuration interfaces, they should be compatible with the existing old configuration. Otherwise a large number of users will be affected.

I think it is sometimes necessary to make breaking changes, so as not to maintain the same interface twice.

**Comment by A2va** [comment]
> customDebugConfig does the same thing, so why add an extra configuration?
You can have multiple configurations for one target, it's not a problem. With the old implementation, that was not possible.

**Comment by waruqi** [comment]
> > We should not remove them. Even if there are new configuration interfaces, they should be compatible with the existing old configuration. Otherwise a large number of users will be affected.
> 
> I think it is sometimes necessary to make breaking changes, so as not to maintain the same interface twice.

I don't think so, every pr should make minimal changes, not massive destructive ones.

Also, I don't see any points of improvement in the new configuration interface.

File: package.json (hunk 6 above)
Comment (A2va) [reply package.json:153]: `onBuild` is called before running the target if `onRunMode` is set to `buildRun`

**Comment by waruqi** [comment]
and debugger does work for me now.

**Comment by A2va** [comment]
> Also, I don't see any points of improvement in the new configuration interface.

This is much more in line with the way other vscode debugging extensions work. 

> and debugger does not work for me now.$

It's weird, I tested it on Windows with cpptools and codelldb and both of them worked.

File: package.json (hunk 9 above)
Comment (A2va) [reply package.json:644]: Ok, I revert back to the old one

**Comment by waruqi** [comment]
Also, now onRun has gone and loaded the debugger too, which it shouldn't have done. We just need to run `xmake run`, which will better support the runtime configuration in xmake.lua.

**Comment by waruqi** [comment]
> This is much more in line with the way other vscode debugging extensions work.

But it doesn't work, and we shouldn't have to break existing configuration behaviour.

image
image

**Comment by waruqi** [comment]
> This is much more in line with the way other vscode debugging extensions work.

I accept patches for native debugging support, but only if they don't break existing configurations and tasks, and provide the smallest changes possible.

Even extraneous code formatting and changes to generic implementations such as bytes2string. Not that they can't be changed completely, of course, but they should be done in a separate pr. The smaller the change, the easier it is to review and merge the patch.

If the patch is too large and heavily broken, it's hard for me to read it and accept the merge.

**Comment by waruqi** [comment]
It also introduces more bugs and doesn't make it easier for me to do testing.

A previous major explorer-related patch introduced a number of bugs that affected a lot of users and is still not fully stable.

**Comment by A2va** [comment]
I understand your concern, and I propose to leave this PR as it is. I will split the modifications into smaller commit/PR, and take into account your comments/criticism.

**Comment by waruqi** [comment]
sure, you can open a new pr and provide a smaller patch.

**Pull request closed without merging**